python/lsst/butler/datastore/posix/datastore.py

```diff
@@ -0,0 +1,233 @@
+        if configFile:
+            self.config.loadFromFile(configFile)
+
+        self.formatterFactory = FormatterFactory()
+        for storageClass, formatter in self.config.formatters.items():
+            self.formatterFactory.registerFormatter(storageClass, formatter)
+
+    def get(self, uri, storageClass, parameters=None):
+        """Load an `InMemoryDataset` from the store.
+
+        Parameters
+        ----------
+        uri : `str`
+            a Universal Resource Identifier that specifies the location of the stored `Dataset`.
+        storageClass : `StorageClass`
+            the `StorageClass` associated with the `DatasetType`.
+        parameters : `dict`
+            `StorageClass`-specific parameters that specify a slice of the `Dataset` to be loaded.
+
+        Returns
+        -------
+        inMemoryDataset : `InMemoryDataset`
+            Requested `Dataset` or slice thereof as an `InMemoryDataset`.
+        """
+        formatter = self.formatterFactory.getFormatter(storageClass)
+        location = self.locationFactory.fromUri(uri)
+        if not os.path.exists(location.path):
+            raise ValueError("No such file: {0}".format(location.uri))
+        return formatter.read(FileDescriptor(location, storageClass.type, parameters))
+
+    def put(self, inMemoryDataset, storageClass, path, typeName=None):
+        """Write a `InMemoryDataset` with a given `StorageClass` to the store.
+
+        Parameters
+        ----------
+        inMemoryDataset : `InMemoryDataset`
+            The `Dataset` to store.
+        storageClass : `StorageClass`
+            The `StorageClass` associated with the `DatasetType`.
+        path : `str`
+            A `Path` that provides a hint that the `Datastore` may use as (part of) the URI.
+        typeName : `str`
+            The `DatasetType` name, which may be used by this `Datastore` to override the
+            default serialization format for the `StorageClass`.
+
+        Returns
+        -------
+        uri : `str` 
+            The `URI` where the primary `Dataset` is stored.
+        components : `dict`, optional
+            A dictionary of URIs for the `Dataset`' components.
+            The latter will be empty if the `Dataset` is not a composite.
+        """
+        formatter = self.formatterFactory.getFormatter(storageClass, typeName)
+        location = self.locationFactory.fromPath(path)
+        return formatter.write(inMemoryDataset, FileDescriptor(location, storageClass.type))
+
+    def remove(self, uri):
+        """Indicate to the Datastore that a `Dataset` can be removed.
+
+        Parameters
+        ----------
+        uri : `str`
+            a Universal Resource Identifier that specifies the location of the stored `Dataset`.
+
+        .. note::
+            Some Datastores may implement this method as a silent no-op to disable `Dataset` deletion through standard interfaces.
+        """
+        location = self.locationFactory.fromUri(uri)
```

Comment:
Isn't there a little more that needs to be done here if the `uri` describes a composite data set?

Reply:
There might be, although in the design we have used so far the fact that `Datastore`s don't know about composites (unless they choose to store one as a single file, in which case matters get tricky). Implementing of composites was outside the scope of this current ticket. Pushed to DM-13033